napari/conftest.py

```diff
@@ -759,15 +879,121 @@ def my_start(self):
+                # so we could provide one that will set object name
+
+                def before_close_func_(w):
+                    w.setObjectName('handled_widget')
+            else:
+                # user provided custom teardown function
+                # so we need to wrap it to set object name
+
+                def before_close_func_(w):
+                    before_close_func(w)
+                    w.setObjectName('handled_widget')
+
+            super().addWidget(widget, before_close_func=before_close_func_)
+
+    @pytest.fixture
+    def qtbot(qapp, request):  # pragma: no cover
+        """
+        Fixture used to create a OwnQtBot instance for using during testing.
+
+        Make sure to call addWidget for each top-level widget you create to ensure
+        that they are properly closed after the test ends.
+
+        The `qapp` fixture is used to ensure that the QApplication is created before,
+        so we need it, even without using it directly in this fixture
+        """
+        return OwnQtBot(request)
+
+
+@pytest.fixture
+def _find_dangling_widgets(request, qtbot):
+    yield
+
+    from qtpy.QtWidgets import QApplication
+
+    from napari._qt.qt_main_window import _QtMainWindow
+
+    top_level_widgets = QApplication.topLevelWidgets()
+
+    viewer_weak_set = getattr(request.node, '_viewer_weak_set', set())
+
+    problematic_widgets = []
+
+    for widget in top_level_widgets:
+        if widget.parent() is not None:
+            continue
+        if (
+            isinstance(widget, _QtMainWindow)
+            and widget._qt_viewer.viewer in viewer_weak_set
+        ):
+            continue
+
+        if widget.__class__.__module__.startswith('qtconsole'):
+            continue
+
+        if widget.objectName() == 'handled_widget':
+            continue
+
+        if widget.__class__.__name__ == 'CanvasBackendDesktop':
+            # leaking in the tests_sys_info.py
+            continue
+
+        problematic_widgets.append(widget)
+
+    if problematic_widgets:
+        text = '\n'.join(
+            f'Widget: {widget} of type {type(widget)} with name {widget.objectName()}'
+            for widget in problematic_widgets
+        )
+        raise RuntimeError(f'Found dangling widgets:\n{text}')
+    for widget in problematic_widgets:
+        widget.setObjectName('handled_widget')
+
+
 def pytest_runtest_setup(item):
+    """Because we have headless test suite we cannot use
+    @pytest.fixture(autouse=True) for fixtures that are detecting
+    leaking Qt objects.
+    So we use here  trick to detect if the `qapp` fixture is used to detect test
+    that are using Qt and need to be checked for Qt objects leaks.
+
+    It may happen that some test is using Qt but is not using `qapp` fixture.
+    This should be caught during the PullRequest review.
+    If you found such a test, please add `qapp` fixture to it.
+    As lack of this fixture may end reporting of a problem in the wrong place.
```

Comment:
Thanks for this note. To clarify: qapp should be used even if qtbot is used?

Reply:
Creating `QtWidget` before creating an instance of `QApplication`. Moreover, `QApplication` is required to have an event loop, that is required for signals. 

But tests do not require it directly. It is enough to be requested by fixture. 

I add comment to prevent accidental removing it in followup PR.